docs/proposals/07-shoot-control-plane-migration.md

```diff
@@ -202,43 +202,99 @@ The ShootState synchronization controller will become part of the gardenlet. It
+When migration is triggered:
+1. The ETCD Druid in the __Destination Seed__ is informed about the ownership of the ETCD cluster and that Control Plane Migration has been triggered
+2. It triggeres `etcd-backup-restore migrate` which first checks to see whether a finished snapshot has been made by the `etcd-backup-restore` sidecar in the __Source Seed__ and then begins the copy operation. If no snapshot is created within a timeout it will copy the latest snapshot.
+4. The `etcd-backup-restore` sidecar initially (during startup) and continuously (when uploading backups) checks to see if a copy operation to a new bucket has started or not.
+    1. If it finds that a copy operation __from__ its bucket is in process or has been finished, it determines that it is in the __Source Seed__ and that control plane migration has been triggered.
+    2. If it finds that a copy operation __to__ its bucket is in process, it determines that it is in the __Destination Seed__ and that control plane migration has been triggered. Once it finds that the copy operation has completed successfully it restores the ETCD data.
+    3. If the ownership information matches and thare is no indication that a copy operation is in process, `etcd-backup-restore` continues to work exactly the way it does today.
+5. When the `etcd-backup-restore` sidecar in the `Shoot's` ETCD pod determines that migration is on and that it is on the __Source Seed__, it stops taking snapshots (full and delta snapshots) and terminates with an error
+    1. This should also cut-off the client requests because the `readinessProbe` of the main ETCD container probes the `etcd-backup-restore` sidecar.
+    1. Just to be sure, the health check in the backup-restore sidecar (used by the main ETCD container's `readinessProbe`) is enhanced to also compare the ownership information passed by `etcd-druid` with what is in the backup. This should ensure that if the main etcd container `readinessProbe` will continue to fail if the `etcd-backup-restore` sidecar comes back up.
+    1. Please note that the etcd container itself might not die here. But the client requests to the container will be cut off.
+5. When the `etcd-backup-restore` sidecar in the `Shoot's` ETCD pod determines that migration is on and that it is on the __Destination Seed__, it first waits for the migration process to finish
+    1. It then checks the available backups and restores the ETCD data from the snapshot with the latest timestamp (older than the migration timestamp).
+7. Once the ETCD container stops accepting the client requests in the __Source Seed__, all components which depend on it (e.g. the Kube APIServer, Machine Controller Manager and so on) will also stop working.
 
-#### If components in the __Source Seed__ cannot reliably read who the leader is from the Garden cluster
 
-Currently we have come up with two ideas to handle this case:
+#### Stopping extension controllers
```

Comment:
I like a lot that we have covered so much ground here along different dimensions:
- Gardener and/or Seed API Server
- Gardenlet, extensions, from PoV of the different components
- Reachability not generally impaired (all-or-nothing), but also dependent on node or load/temporal 
- Reconciliation/operation still ongoing or not

What I am wondering however: wasn't there a good way to propagate what the `etcd-backup-restore` sidecar knows (when it sees that migration was triggered) to the rest of the bunch (`gardenlet` and by that also the extensions) or was that considered dirty?
A very long time ago, we thought, the safest way to coordinate a hand-shake is by revoking access (difficult) or DNS (TTLs and cache implementation may be tricky/unsafe, though) or a blob store and the one the sidecar has access to is ideal, because it's less likely to be impaired, it's used by the ETCD pod already and that's the most critical resource anyway that we want to inform of the migration (so that ETCD can shut down and the final backup can be taken).

Reply:
Yes indeed we discussed that. For that to work we would need something - ETCD Druid seems like the best fit, to monitor the health of the `etcd-backup-restore` sidecar, and then update a "lease" resource which the other controllers read. But we are more inclined to leaving the `gardenlet` juts checking if it can connect to the Garden APIServer and if it can - read the leader info from the `Shoot` spec and update this "lease".